(91) Show optional fields and validation errors

leeky opened the pull request:
- Forms now indicate when fields are optional rather than when they are required
 - The 'describe unknown repair' and 'contact details' forms have required fields, so now have form objects to perform validation
 - Validation prevents incomplete forms being submitted
 - Validation errors are in the GOVUK style
 -  Optional fields need to have labels

<img width="517" alt="validation-errors" src="https://user-images.githubusercontent.com/3166/31181406-23af8be6-a919-11e7-9ed5-6260587f2fc5.png">

NB: GOVUK style suggested having a series of links to the invalid fields at the top of the page. This is quite hard to do because it requires some additional ARIA attributes on the form and error fields. I've therefore omitted this for now, which feels fine since we're avoiding too many questions per page.

leeky committed:
(feature) Optional rather than required fields are marked

Simple Form defaults to marking required fields with an asterisk,
whereas GOVUK (and others) recommend marking optional fields instead.

The fix is heavily inspired by this blog post:
https://jeremysmith.co/posts/2015-12-03-how-to-mark-optional-form-fields-with-simple-form/

leeky committed:
(chore) Correct classes applied by SimpleForm to errors

This allows invalid form elements to get the correct styling from the
GOVUK toolkit.

Also, defaulted buttons to have the class of `button` to save typing!

leeky committed:
(feature) Add validation to contact details & unknown repair forms

In order to make the fields contained in these forms required, a form
object was added to store the validation rules.

The `submit` actions now both check that the forms are valid before
allowing the user to proceed.

leeky committed:
(fix) Optional form fields need to have labels

Now that optional fields are suffixed with '(optional)', those fields
need to have labels.

leeky committed:
(chore) Add missing translations

NB: Adding these under the `activemodel` key in `en.yml` means that they
used in the validation errors, as well for the form field labels.

I've included `callback_time` in both places, because the wording needs
to be different depending on context:

  - Form field should ask "What time is best to call you?"
  - Validation error is "Best time to call you", to avoid the strange
    "What time is best to call you? can't be blank"

dgmstuart commented on spec/models/contact_details_form_spec.rb:
```diff
@@ -0,0 +1,26 @@
+
+RSpec.describe ContactDetailsForm do
+  describe 'validations' do
+    it '.full_name' do
```
I guess 'it validates presence of full name' is more descriptive but more verbose?
This is fine for now, but at some point we're going to need additional validations - e.g. field length

dgmstuart commented on spec/models/contact_details_form_spec.rb:
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+RSpec.describe ContactDetailsForm do
+  describe 'validations' do
```
Do we need a spec that shows that this is valid when all fields are present? I think currently this would pass if `valid?` was always false.

leeky replied:
:+1: Added

dgmstuart commented on spec/models/contact_details_form_spec.rb:
```diff
@@ -0,0 +1,26 @@
+      expect(form.errors.details[:telephone_number]).to include(error: :blank)
+    end
+
+    it '.callback_time' do
```
Would there be value in demonstrating that this form is valid when only one checkbox is checked?

dgmstuart commented on config/initializers/simple_form.rb:
```diff
@@ -47,4 +52,7 @@
   config.boolean_label_class = 'checkbox'
+
+  # Append 'optional' label instead of prepending
+  config.label_text = ->(label, required, _explicit_label) { "#{label} #{required}" }
```
Are we able to rename this parameter to `optional`? Would that make any sense?

dgmstuart replied:
Also: this will result in a space after the label for required fields. Probably no big deal?
It made sense in the blog post because they were marking both required and optional fields.

leeky replied:
I think the default behaviour (prefixing an asterisk to required fields) also adds an extra space. I've fixed this in this instance however.

Also, same with renaming parameter to `optional`.

dgmstuart commented on spec/features/users_cannot_submit_incomplete_forms_spec.rb:
```diff
@@ -0,0 +1,59 @@
+    within '.contact_details_form_callback_time' do
+      expect(page).to have_css('span.error-message')
+      expect(page).to have_css('div input[aria-invalid=true]')
+    end
```
Do we now need further steps in this spec to show that after these validation errors we can go on to successfully submit a form?

dgmstuart commented on app/controllers/contact_details_controller.rb:
```diff
@@ -2,9 +2,23 @@ class ContactDetailsController < ApplicationController
 
   def submit
+    selected_answer_store = SelectedAnswerStore.new(session)
+    @selected_answers = selected_answer_store.selected_answers['address']
```
Do we need @selected_answers in the happy path case?
I'm never quite sure if it's worth making the distinction in this sort of situation.

leeky replied:
I did this so that the `index` and `submit` methods were similar shaped, rather than having the `SelectedAnswerStore` being in the conditional.

dgmstuart commented:
The validations on the first page don't seem to be styled (yes, I had to force the submit button to be displayed using the inspector):
<img width="663" alt="screen shot 2017-10-04 at 16 52 14" src="https://user-images.githubusercontent.com/393167/31185670-7da588c0-a924-11e7-9a29-43e6a2f0bd06.png">

dgmstuart commented:
Ditto the Address selection page, which also jumps back to the address search page instead of re-rendering the address selection list:

<img width="572" alt="screen shot 2017-10-04 at 16 54 37" src="https://user-images.githubusercontent.com/393167/31185793-c90fa232-a924-11e7-83d6-40992c4c3a74.png">

dgmstuart reviewed (changes requested):
Generally looks great - the errors on individual fields look really nice.

The main thing missing is specs which test the forms being submitted successfully.

leeky committed:
(chore) Start form shows error when submitted without answer

Previously, I used a custom validator which set the error message on
`@form.errors[:base]`, which feels odd as the other forms use standard
attribute validations.

To resolve this, I've renamed the sole form attribute to `answer`, which
means Rails automatically provides "Answer cannot be blank" by default.

It also has the benefit of showing the error inline, with the field
highlighted correctly and the correct ARIA attributes applied.

leeky committed:
(chore) User gets an error if they don't select an address

If the user doesn't select an address, show them an error and redisplay
the address search results.

To achieve this, first the entered postcode is stored in a hidden field
so it can be reused if an error occurs.

The purpose of the `Address` model wasn't particularly clear, so I've
renamed it to `AddressForm`. This change means that the validation of
the required field `property_preference` can be moved into it.

Additionally, when re-rendering the search results within
`AddressesController`, we can use the form object rather than
interrogating `params` directly.

dgmstuart commented on app/controllers/addresses_controller.rb:
```diff
@@ -1,18 +1,20 @@
-        alert: t('addresses.errors.blank')
-      )
-    end
+    @form = AddressForm.new(address_form_params)
```
Nice - I like this :)

dgmstuart commented on app/controllers/addresses_controller.rb:
```diff
@@ -21,11 +23,7 @@ def create
-
-  def property_reference
-    update_params[:property_reference]
+  def address_form_params
```
Ah this tidy up feels good

dgmstuart reviewed (approved):
Looks great to me 👍
